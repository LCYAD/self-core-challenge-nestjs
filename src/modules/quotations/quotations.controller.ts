import {
  Body,
  Controller,
  Get,
  Param,
  ParseArrayPipe,
  Post
} from '@nestjs/common'
import { validationExceptionFactory } from '../../utils/validation.util'
import { QuotationIdParamDto } from '../../dtos/quotation/id.param.dto'
import { QuotationsPostReqDto } from './dtos/quotationsPostReq.dto'
import { QuotationHandlerService } from './quotations.handler.service'

@Controller('quotations')
export class QuotationsController {
  constructor(
    private readonly quotationHandlerService: QuotationHandlerService
  ) {}

  @Get(':id')
  getQuotation(@Param() param: QuotationIdParamDto) {
    return this.quotationHandlerService.getQuotationById(param)
  }

  @Post()
  postQuotations(
    @Body(
      new ParseArrayPipe({
        items: QuotationsPostReqDto,
        exceptionFactory: validationExceptionFactory
      })
    )
    body: QuotationsPostReqDto[]
  ) {
    return this.quotationHandlerService.createQuotations(body)
  }
}
