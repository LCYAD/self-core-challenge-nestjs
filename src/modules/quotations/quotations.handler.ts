import { Injectable } from '@nestjs/common'

import _ from 'lodash'
import { customAlphabet } from 'nanoid'

import type { QuotationBaseDto } from '@dtos/quotation/base.dto'
import { QuotationIdParamDto } from '@dtos/quotation/id.param.dto'
import { ResourcesNotFoundException } from '@exceptions/resourcesNotFound.exception'
import { redis } from '@utils/redis.util'

import type { QuotationsPostReqDto } from './dtos/quotationsPostReq.dto'

@Injectable()
export class QuotationsHandler {
  async getQuotationById({ id }: QuotationIdParamDto) {
    const quotation = await redis.get(id)
    if (!quotation) {
      throw new ResourcesNotFoundException(
        'QuotationsHandler:getQuotationById',
        'Quotation not found'
      )
    }
    return JSON.parse(quotation)
  }

  async createQuotations(body: QuotationsPostReqDto[]) {
    const now = Date.now()
    const duration = 5 * 60 * 1000
    const expireAt = new Date(now + duration).toISOString()
    const generateQuotationId = () => customAlphabet('0123456789', 12)()
    const quotations = _.map(body, (quotation) => ({
      ...quotation,
      quotationId: generateQuotationId(),
      expireAt
    }))
    await Promise.all(
      _.map(quotations, (quotation: QuotationBaseDto) =>
        redis.set(quotation.quotationId, JSON.stringify(quotation))
      )
    )
    return quotations
  }
}
