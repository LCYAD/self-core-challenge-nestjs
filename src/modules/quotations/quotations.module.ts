import { Module } from '@nestjs/common'
import { QuotationsController } from './quotations.controller'
import { QuotationHandlerService } from './quotations.handler.service'

@Module({
  imports: [],
  controllers: [QuotationsController],
  providers: [QuotationHandlerService]
})
export class QuotationsModule {}
