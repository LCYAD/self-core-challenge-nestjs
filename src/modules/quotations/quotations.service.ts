import { Injectable } from '@nestjs/common'
import { QUOTATION } from '../../constants/fakers/quotation'

@Injectable()
export class QuotationHandlerService {
  getQuotationById() {
    return QUOTATION
  }
}
