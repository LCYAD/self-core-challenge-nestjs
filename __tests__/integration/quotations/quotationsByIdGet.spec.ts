import { HttpStatus } from '@nestjs/common'
import {
  FastifyAdapter,
  NestFastifyApplication
} from '@nestjs/platform-fastify'

import { COMMON_FIELD } from '@constants/fakers/common.faker.constant'
import { quotationBaseValidExample } from '@dtos/quotation/base.dto'
import { QuotationsHandler } from '@modules/quotations/quotations.handler'
import { QuotationsModule } from '@modules/quotations/quotations.module'
import {
  generateSequentialTests,
  mockGenericApp
} from '@tests/helper/e2e.helper'
import { redis } from '@utils/redis.util'

jest.mock('ioredis', () => require('ioredis-mock'))

const path = `/quotations/${COMMON_FIELD.QUOTATION_ID_FIELD}`

describe('GET /quotations/:id', () => {
  let app: NestFastifyApplication
  let quotationsHandler: QuotationsHandler
  let mockFunctionGroup

  beforeAll(async () => {
    const appModule: NestFastifyApplication = (
      await mockGenericApp({
        module: {
          imports: [QuotationsModule]
        }
      })
    ).createNestApplication(new FastifyAdapter())

    quotationsHandler = appModule.get(QuotationsHandler)
    mockFunctionGroup = {
      spyHandlerGetQuotationById: jest
        .spyOn(quotationsHandler, 'getQuotationById')
        .mockName('spyHandlerGetQuotationById'),
      spyRedisGet: jest.spyOn(redis, 'get')
    }

    app = await appModule.init()
  })

  describe('success cases', () => {
    it('should return correct response', async () => {
      mockFunctionGroup.spyRedisGet.mockReturnValue(
        JSON.stringify(quotationBaseValidExample)
      )
      const response = await app.inject({
        method: 'GET',
        url: path
      })
      expect(response.statusCode).toEqual(HttpStatus.OK)
      generateSequentialTests({
        mockFunctionGroup,
        snapshots: ['spyHandlerGetQuotationById', 'spyRedisGet'],
        apiResponse: response
      })
    })
  })

  describe('error cases', () => {
    it('should throw Validation Error if quotation id is invalid', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/quotations/xxx'
      })
      expect(response.statusCode).toEqual(HttpStatus.UNPROCESSABLE_ENTITY)
      generateSequentialTests({
        mockFunctionGroup,
        notCalled: ['spyHandlerGetQuotationById', 'spyRedisGet'],
        apiResponse: response
      })
    })
    it('should throw error if quotation is not found', async () => {
      mockFunctionGroup.spyRedisGet.mockReturnValue()
      const response = await app.inject({
        method: 'GET',
        url: path
      })
      expect(response.statusCode).toEqual(HttpStatus.NOT_FOUND)
      generateSequentialTests({
        mockFunctionGroup,
        snapshots: ['spyHandlerGetQuotationById', 'spyRedisGet'],
        apiResponse: response
      })
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(async () => {
    await app.close()
  })
})
