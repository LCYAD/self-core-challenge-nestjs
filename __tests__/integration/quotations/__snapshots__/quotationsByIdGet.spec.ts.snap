// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /quotations/:id error cases should throw Validation Error if quotation id is invalid 1`] = `
{
  "code": "10001",
  "detail": [
    {
      "msg": "{"matches":"id must match /^[0-9]{12}$/ regular expression"}",
      "path": "id",
      "value": "xxx",
    },
  ],
  "type": "VALIDATION_ERROR",
}
`;

exports[`GET /quotations/:id error cases should throw error if quotation is not found 1`] = `
[MockFunction spyHandlerGetQuotationById] {
  "calls": [
    [
      QuotationIdParamDto {
        "id": "182379172831",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /quotations/:id error cases should throw error if quotation is not found 2`] = `
[MockFunction] {
  "calls": [
    [
      "182379172831",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GET /quotations/:id error cases should throw error if quotation is not found 3`] = `
{
  "code": "10004",
  "detail": "Quotation not found",
  "type": "RESOURCES_NOT_FOUND",
}
`;

exports[`GET /quotations/:id success cases should return correct response 1`] = `
[MockFunction spyHandlerGetQuotationById] {
  "calls": [
    [
      QuotationIdParamDto {
        "id": "182379172831",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /quotations/:id success cases should return correct response 2`] = `
[MockFunction] {
  "calls": [
    [
      "182379172831",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "{"quotationId":"182379172831","expireAt":"2022-05-21T00:05:00.000Z","scheduleAt":"2022-05-21T00:05:00.000Z","deliveryBy":"2022-05-21T01:00:00.000Z","stops":[{"coordinates":{"lat":"22.335313","lng":"114.175840"},"address":"999 Prince Edward Road"},{"coordinates":{"lat":"22.335313","lng":"114.175840"},"address":"InnoCentre, 72 Tat Chee Avenue"}],"location":"Hong Kong","item":{"quantity":1,"weight":1,"categories":["test"],"handlingInstructions":["test"]}}",
    },
  ],
}
`;

exports[`GET /quotations/:id success cases should return correct response 3`] = `
{
  "deliveryBy": "2022-05-21T01:00:00.000Z",
  "expireAt": "2022-05-21T00:05:00.000Z",
  "item": {
    "categories": [
      "test",
    ],
    "handlingInstructions": [
      "test",
    ],
    "quantity": 1,
    "weight": 1,
  },
  "location": "Hong Kong",
  "quotationId": "182379172831",
  "scheduleAt": "2022-05-21T00:05:00.000Z",
  "stops": [
    {
      "address": "999 Prince Edward Road",
      "coordinates": {
        "lat": "22.335313",
        "lng": "114.175840",
      },
    },
    {
      "address": "InnoCentre, 72 Tat Chee Avenue",
      "coordinates": {
        "lat": "22.335313",
        "lng": "114.175840",
      },
    },
  ],
}
`;
